### PR TITLE
Merge concurrent API requests.

### PR DESCRIPTION
**Pull request by mykhailo-kuchma**
Exact same API requests are blocked, and wait for the first one to
complete.

Relates-To: OLPEDGE-2040

Signed-off-by: Mykhailo Kuchma <ext-mykhailo.kuchma@here.com>

### CODE REVIEW

File: olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.cpp [hunk 1]
```diff
@@ -149,6 +136,14 @@ DataResponse DataRepository::GetBlobData(
     return {{client::ErrorCode::PreconditionFailed, "Data handle is missing"}};
   }
 
+  NamedMutex mutex(catalog.ToString() + *data_handle);
+  std::unique_lock<NamedMutex> lock(mutex, std::defer_lock);
+
```
Comment (kzvieriev) [inline olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.cpp:141]: Maybe add layer to the mutex name.

File: olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.cpp [hunk 2]
```diff
@@ -93,20 +94,6 @@ DataResponse DataRepository::GetVersionedData(
...
-      version_request.WithFetchOption(request.GetFetchOption())
-          .WithBillingTag(request.GetBillingTag());
-      auto latest_version_response =
-          repository::CatalogRepository::GetLatestVersion(
-              catalog, context, std::move(version_request), settings);
-      if (!latest_version_response.IsSuccessful()) {
-        return latest_version_response.GetError();
-      }
-      request.WithVersion(latest_version_response.GetResult().GetVersion());
-    }
```
Comment (kzvieriev) [inline olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.cpp:108]: DataRequest::GetVersion is deprecated, but I think we can't just remove this block.
GetPartitionById() used below expecting for a version in the request, so this change breaking DataRepository::GetVersionedData() logic.
Is this related to merging requests?

File: olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.cpp (hunk 1 above)
Comment (andescu) [reply olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.cpp:141]: Good point. According to DS documentation, the data handle is only to be unique per layer.

Comment (mykhailo-kuchma) [reply olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.cpp]: Added layer

File: olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.cpp (hunk 2 above)
Comment (mykhailo-kuchma) [reply olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.cpp]: Actually we don't enter this branch. But I reverted this code for now.

File: olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp [hunk 3]
```diff
@@ -367,6 +367,18 @@ PartitionResponse PartitionsRepository::GetAggregatedTile(
     const client::OlpClientSettings& settings) {
   const auto fetch_option = request.GetFetchOption();
   const auto& tile_key = request.GetTileKey();
+
+  const auto& root_tile_key = tile_key.ChangedLevelBy(-kAggregateQuadTreeDepth);
+  const auto root_tile_here = root_tile_key.ToHereTile();
+
+  NamedMutex mutex(catalog.ToString() + layer + root_tile_here + "Index");
+  std::unique_lock<NamedMutex> lock(mutex, std::defer_lock);
```
Comment (kzvieriev) [inline olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp:375]: This is tricky one :-)
GetAggregatedTile is looking in cache for any tree containing it, not only `tile_key.ChangedLevelBy(-4)`.
So we need to block loading of quad trees from root_tile_key.Level() to tile_key.Level(). Not sure this can be done using just string.

Comment (mykhailo-kuchma) [reply olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp:375]: It's needed. Since this is the tile that will be used for API call. Our goal is to merge the online requests.
